mitmproxy/addons/session.py

```diff
@@ -67,7 +111,278 @@ def is_session_db(path):
+            for row in rows:
+                flow = protobuf.loads(row[0])
+                if row[1]:
+                    typ = self.type_mappings["body"][row[1]]
+                    if typ and row[2]:
+                        setattr(getattr(flow, typ), "content", row[2])
+                flow = self._reassemble(flow)
+                flows.append(flow)
+        return flows
+
+    def clear(self):
+        self.con.executescript("DELETE FROM body; DELETE FROM annotation; DELETE FROM flow;")
+
+
+matchall = flowfilter.parse(".")
+
+orders = [
+    "time",
+    "method",
+    "url",
+    "size"
+]
+
+
+class Session:
+    def __init__(self):
+        self.db_store = None
+        self._hot_store = collections.OrderedDict()
+        self._live_components = {}
+        self._view = []
+        self.order = orders[0]
+        self.filter = matchall
+        self._flush_period = 3.0
+        self._tweak_period = 0.5
+        self._flush_rate = 150
+        self.started = False
+
+    def load(self, loader):
+        loader.add_option(
+            "session_path", typing.Optional[types.Path], None,
+            "Path of session to load or to create."
+        )
+        loader.add_option(
+            "view_order", str, "time",
+            "Flow sort order.",
+            choices=list(map(lambda c: c[1], orders))
+        )
+        loader.add_option(
+            "view_filter", typing.Optional[str], None,
+            "Limit the view to matching flows."
+        )
+
+    def running(self):
+        if not self.started:
+            self.started = True
+            self.db_store = SessionDB(ctx.options.session_path)
+            loop = asyncio.get_event_loop()
+            loop.create_task(self._writer())
+            loop.create_task(self._tweaker())
+
+    def configure(self, updated):
+        if "view_order" in updated:
+            self.set_order(ctx.options.view_order)
+        if "view_filter" in updated:
+            self.set_filter(ctx.options.view_filter)
+
+    async def _writer(self):
+        while True:
+            await asyncio.sleep(self._flush_period)
+            tof = []
+            to_dump = min(self._flush_rate, len(self._hot_store))
+            for _ in range(to_dump):
+                tof.append(self._hot_store.popitem(last=False)[1])
+            self.db_store.store_flows(tof)
+
+    async def _tweaker(self):
+        while True:
+            await asyncio.sleep(self._tweak_period)
+            if len(self._hot_store) >= 3 * self._flush_rate:
+                self._flush_period *= 0.9
+                self._flush_rate *= 1.1
+            elif len(self._hot_store) < self._flush_rate:
+                self._flush_period *= 1.1
+                self._flush_rate *= 0.9
+
+    def load_view(self):
+        ids = [fid for _, fid in self._view]
+        flows = self.load_storage(ids)
+        return sorted(flows, key=lambda f: self._generate_order(f))
+
+    def load_storage(self, ids=None):
```

Comment:
Return type annotation for this function, please. Does this return a list of flows? Or a list of  `Optional[flow]`?